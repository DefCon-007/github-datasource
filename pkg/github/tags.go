package github

import (
	"context"
	"time"

	"github.com/grafana/grafana-github-datasource/pkg/models"
	"github.com/grafana/grafana-plugin-sdk-go/data"
	"github.com/shurcooL/githubv4"
)

// Tag is a GitHub tag. Every tag has an associated commit
type Tag struct {
	Name   string
	Target struct {
		Commit Commit `graphql:"... on Commit"`
	}
}

// Tags is a list of GitHub tags
type Tags []Tag

// Frames converts the list of tags to a Grafana DataFrame
func (t Tags) Frames() data.Frames {
	frame := data.NewFrame(
		"tags",
		data.NewField("name", nil, []string{}),
		data.NewField("id", nil, []string{}),
		data.NewField("author", nil, []string{}),
		data.NewField("author_login", nil, []string{}),
		data.NewField("author_email", nil, []string{}),
		data.NewField("author_company", nil, []string{}),
		data.NewField("commited_at", nil, []time.Time{}),
		data.NewField("pushed_at", nil, []time.Time{}),
	)

	for _, v := range t {
		frame.AppendRow(
			v.Name,
			v.Target.Commit.OID,
			v.Target.Commit.Author.Name,
			v.Target.Commit.Author.User.Login,
			v.Target.Commit.Author.Email,
			v.Target.Commit.Author.User.Company,
			v.Target.Commit.CommittedDate.Time,
			v.Target.Commit.PushedDate.Time,
		)
	}

	return data.Frames{frame}
}

// QueryListTags is the GraphQL query for listing GitHub tags in a repository
// {
//   repository(name: "grafana", owner: "grafana") {
//     refs(refPrefix: "refs/tags/", first: 100) {
//       nodes {
//         name
//         target {
//           ... on Commit {
//             message
//             pushedDate
//             author {
//               name
//               email
//             }
//           }
//         }
//       }
//     }
//   }
// }
type QueryListTags struct {
	Repository struct {
		Refs struct {
			Nodes    []Tag
			PageInfo PageInfo
		} `graphql:"refs(refPrefix: \"refs/tags/\", first: 100, after: $cursor)"`
	} `graphql:"repository(name: $name, owner: $owner)"`
}

// GetAllTags retrieves every tag from a repository
func GetAllTags(ctx context.Context, client Client, opts models.ListTagsOptions) (Tags, error) {
	var (
		variables = map[string]interface{}{
			"cursor": (*githubv4.String)(nil),
			"owner":  githubv4.String(opts.Owner),
			"name":   githubv4.String(opts.Repository),
		}

		tags = []Tag{}
	)

	for {
		q := &QueryListTags{}
		if err := client.Query(ctx, q, variables); err != nil {
			return nil, err
		}
		tags = append(tags, q.Repository.Refs.Nodes...)
		if !q.Repository.Refs.PageInfo.HasNextPage {
			break
		}
		variables["cursor"] = q.Repository.Refs.PageInfo.EndCursor
	}

	return tags, nil
}

// GetTagsInRange retrieves every tag from the repository and then returns the ones that fall within the given time range.
func GetTagsInRange(ctx context.Context, client Client, opts models.ListTagsOptions, from time.Time, to time.Time) (Tags, error) {
	tags, err := GetAllTags(ctx, client, opts)
	if err != nil {
		return nil, err
	}

	filtered := []Tag{}

	for i, v := range tags {
		if v.Target.Commit.CommittedDate.After(from) && v.Target.Commit.CommittedDate.Before(to) {
			filtered = append(filtered, tags[i])
		}
	}

	return filtered, nil
}
