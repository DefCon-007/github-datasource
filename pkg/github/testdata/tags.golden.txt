ðŸŒŸ This was machine generated.  Do not edit. ðŸŒŸ

Frame[0] 
Name: tags
Dimensions: 8 Fields by 2 Rows
+----------------+----------------+----------------+--------------------+--------------------+----------------------+---------------------------------+---------------------------------+
| Name: name     | Name: id       | Name: author   | Name: author_login | Name: author_email | Name: author_company | Name: commited_at               | Name: pushed_at                 |
| Labels:        | Labels:        | Labels:        | Labels:            | Labels:            | Labels:              | Labels:                         | Labels:                         |
| Type: []string | Type: []string | Type: []string | Type: []string     | Type: []string     | Type: []string       | Type: []time.Time               | Type: []time.Time               |
+----------------+----------------+----------------+--------------------+--------------------+----------------------+---------------------------------+---------------------------------+
|                |                | firstCommitter | firstCommitter     | first@example.com  | ACME Corp            | 2020-08-25 16:21:56 +0000 +0000 | 2020-08-25 16:23:56 +0000 +0000 |
|                |                | firstCommitter | firstCommitter     | first@example.com  | ACME Corp            | 2020-08-25 17:21:56 +0000 +0000 | 2020-08-25 18:21:56 +0000 +0000 |
+----------------+----------------+----------------+--------------------+--------------------+----------------------+---------------------------------+---------------------------------+


====== TEST DATA RESPONSE (arrow base64) ======
FRAME=QVJST1cxAAD/////0AMAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAFQAAAACAAAAKAAAAAQAAAC0/P//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAANT8//8IAAAAEAAAAAQAAAB0YWdzAAAAAAQAAABuYW1lAAAAAAgAAADkAgAAfAIAACACAAC0AQAASAEAANwAAABsAAAABAAAAE79//8UAAAAQAAAAEAAAAAAAAAKQAAAAAEAAAAEAAAAPP3//wgAAAAUAAAACQAAAHB1c2hlZF9hdAAAAAQAAABuYW1lAAAAAAAAAACa////AAADAAkAAABwdXNoZWRfYXQAAACy/f//FAAAAEAAAABIAAAAAAAACkgAAAABAAAABAAAAKD9//8IAAAAFAAAAAsAAABjb21taXRlZF9hdAAEAAAAbmFtZQAAAAAAAAAAAAAGAAgABgAGAAAAAAADAAsAAABjb21taXRlZF9hdAAe/v//FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAAAz+//8IAAAAGAAAAA4AAABhdXRob3JfY29tcGFueQAABAAAAG5hbWUAAAAAAAAAAAz+//8OAAAAYXV0aG9yX2NvbXBhbnkAAIb+//8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAAdP7//wgAAAAYAAAADAAAAGF1dGhvcl9lbWFpbAAAAAAEAAAAbmFtZQAAAAAAAAAAdP7//wwAAABhdXRob3JfZW1haWwAAAAA7v7//xQAAABEAAAARAAAAAAAAAVAAAAAAQAAAAQAAADc/v//CAAAABgAAAAMAAAAYXV0aG9yX2xvZ2luAAAAAAQAAABuYW1lAAAAAAAAAADc/v//DAAAAGF1dGhvcl9sb2dpbgAAAABW////FAAAADwAAAA8AAAAAAAABTgAAAABAAAABAAAAET///8IAAAAEAAAAAYAAABhdXRob3IAAAQAAABuYW1lAAAAAAAAAAA8////BgAAAGF1dGhvcgAArv///xQAAAA4AAAAOAAAAAAAAAU0AAAAAQAAAAQAAACc////CAAAAAwAAAACAAAAaWQAAAQAAABuYW1lAAAAAAAAAACQ////AgAAAGlkAAAAABIAGAAUAAAAEwAMAAAACAAEABIAAAAUAAAARAAAAEgAAAAAAAAFRAAAAAEAAAAMAAAACAAMAAgABAAIAAAACAAAABAAAAAEAAAAbmFtZQAAAAAEAAAAbmFtZQAAAAAAAAAABAAEAAQAAAAEAAAAbmFtZQAAAAD/////OAIAABQAAAAAAAAADAAWABQAEwAMAAQADAAAAAABAAAAAAAAFAAAAAAAAAMDAAoAGAAMAAgABAAKAAAAFAAAAHgBAAACAAAAAAAAAAAAAAAWAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAEAAAAAAAAAAwAAAAAAAAACAAAAAAAAAAUAAAAAAAAAAAAAAAAAAAAFAAAAAAAAAAEAAAAAAAAABgAAAAAAAAACAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAEAAAAAAAAACQAAAAAAAAACgAAAAAAAAAuAAAAAAAAAAAAAAAAAAAALgAAAAAAAAAEAAAAAAAAADIAAAAAAAAABgAAAAAAAAA4AAAAAAAAAAAAAAAAAAAAOAAAAAAAAAAEAAAAAAAAADwAAAAAAAAAAAAAAAAAAAA8AAAAAAAAAAQAAAAAAAAAAAAAAAIAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOAAAAHAAAAAAAAABmaXJzdENvbW1pdHRlcmZpcnN0Q29tbWl0dGVyAAAAAAAAAAAOAAAAHAAAAAAAAABmaXJzdENvbW1pdHRlcmZpcnN0Q29tbWl0dGVyAAAAAAAAAAARAAAAIgAAAAAAAABmaXJzdEBleGFtcGxlLmNvbWZpcnN0QGV4YW1wbGUuY29tAAAAAAAAAAAAAAkAAAASAAAAAAAAAEFDTUUgQ29ycEFDTUUgQ29ycAAAAAAAAABo7bJVjy4WAAim45uSLhYAGHyjcY8uFgCoXhTilS4WEAAAAAwAFAASAAwACAAEAAwAAAAQAAAALAAAADwAAAAAAAMAAQAAAOADAAAAAAAAQAIAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAACgAMAAAACAAEAAoAAAAIAAAAVAAAAAIAAAAoAAAABAAAALT8//8IAAAADAAAAAAAAAAAAAAABQAAAHJlZklkAAAA1Pz//wgAAAAQAAAABAAAAHRhZ3MAAAAABAAAAG5hbWUAAAAACAAAAOQCAAB8AgAAIAIAALQBAABIAQAA3AAAAGwAAAAEAAAATv3//xQAAABAAAAAQAAAAAAAAApAAAAAAQAAAAQAAAA8/f//CAAAABQAAAAJAAAAcHVzaGVkX2F0AAAABAAAAG5hbWUAAAAAAAAAAJr///8AAAMACQAAAHB1c2hlZF9hdAAAALL9//8UAAAAQAAAAEgAAAAAAAAKSAAAAAEAAAAEAAAAoP3//wgAAAAUAAAACwAAAGNvbW1pdGVkX2F0AAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMACwAAAGNvbW1pdGVkX2F0AB7+//8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAADP7//wgAAAAYAAAADgAAAGF1dGhvcl9jb21wYW55AAAEAAAAbmFtZQAAAAAAAAAADP7//w4AAABhdXRob3JfY29tcGFueQAAhv7//xQAAABEAAAARAAAAAAAAAVAAAAAAQAAAAQAAAB0/v//CAAAABgAAAAMAAAAYXV0aG9yX2VtYWlsAAAAAAQAAABuYW1lAAAAAAAAAAB0/v//DAAAAGF1dGhvcl9lbWFpbAAAAADu/v//FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAANz+//8IAAAAGAAAAAwAAABhdXRob3JfbG9naW4AAAAABAAAAG5hbWUAAAAAAAAAANz+//8MAAAAYXV0aG9yX2xvZ2luAAAAAFb///8UAAAAPAAAADwAAAAAAAAFOAAAAAEAAAAEAAAARP///wgAAAAQAAAABgAAAGF1dGhvcgAABAAAAG5hbWUAAAAAAAAAADz///8GAAAAYXV0aG9yAACu////FAAAADgAAAA4AAAAAAAABTQAAAABAAAABAAAAJz///8IAAAADAAAAAIAAABpZAAABAAAAG5hbWUAAAAAAAAAAJD///8CAAAAaWQAAAAAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAABEAAAASAAAAAAAAAVEAAAAAQAAAAwAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAQAAABuYW1lAAAAAAQAAABuYW1lAAAAAAAAAAAEAAQABAAAAAQAAABuYW1lAAAAAAAEAABBUlJPVzE=
